### PR TITLE
CMIP6 Refine Diag Script

### PR DESCRIPTION
**Pull request by jkrasting**
@adcroft, @StephenGriffies -

I opened a PR so that we can discuss this development via GitHub.  I'll update this branch as needed and we can close out the PR once we are all happy.

Work in progress ...

### CODE REVIEW

**Comment by adcroft** [comment]
Status of the masking for this PR?

**Comment by jkrasting** [comment]
You mean the topography masking?

**Comment by adcroft** [comment]
Yes - I mean is this the final version where the topography masking works
as we discussed?

--
Dr Alistair Adcroft    (Alistair.Adcroft@noaa.gov)
Princeton University           Tel: (609) 987-5073
NOAA/GFDL, 201 Forrestal Road, Princeton, NJ 08540

**Comment by jkrasting** [comment]
It looks reasonable to me.  Can we keep this PR open for discussion until
we have all of the offline diagnostic scripts complete?  The GitHub code
review feature is helpful.

Still on the docket are the heat transports and the passage flows ('mfo').

**Comment by ashao** [comment]
@jkrasting - That's correct. All that it needs are the netcdf variable handles to umo and vmo.

**Comment by adcroft** [comment]
I gather we might still have issues for the scripts? Let me know when we're ready to pull this PR in.

This PR has changes to the diag_table that look like the ones in #179 (which might need to be corrected for umo). Fix #179 so that we can pull it in and then we can remove the diag_table changes from this PR.

**Comment by jkrasting** [comment]
Yes, @ashao is working on the wmo diagnostic and I am cleaning up the cell
methods attributes.  Other than that, we are almost ready to go.  The
plumbing is all connected and tested in the workflow.

How about I just retract #179 and I give you an all-encompassing PR later
today?

**Comment by adcroft** [comment]
The diag_table in this PR looks right so, yes, go ahead and close #179.

The scripts look good but where do we stand with the `wo` results?

For an east-west section:
  vmo, cell_methods:"z:sum y:point x:sum"
  thetao, cell_methods:"z:mean y:point x:mean"
  vo, cell_methods:"z:mean y:point x:mean"

cell_measures are a bit more complicated because volcello is a volume which doesn't apply to a section. LIkewise for areacello, so I'd leave them out.

**Comment by jkrasting** [comment]
Our sections are N-S, however.   For msftyyz, I have: 'z_i:sum yq:sum
time:mean'  Does that make sense?  Confidence low on this ...

File: tools/analysis/refineDiag_ocean_month.py [hunk 1]
```diff
@@ -119,6 +119,7 @@ def main(args):
     hfbasin = np.ma.array(hfbasin,fill_value=nc_misval)
     hfbasin.long_name = 'Northward Ocean Heat Transport'
     hfbasin.units = 'W'
+    hfbasin.coordinates = 'region'
```
Comment (adcroft) [inline tools/analysis/refineDiag_ocean_month.py]: That looks right to me, based on http://clipc-services.ceda.ac.uk/dreq/u/dim:basin.html.

Can you post a "ncdump -h" of a final file?

Comment (adcroft) [reply tools/analysis/refineDiag_ocean_month.py]: And it also looks right based on http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#geographic-regions .

**Comment by jkrasting** [comment]
Thanks. What about the cell_methods, though?  I'm running the latest codes
now. Will post the ncdump -h soon.

**Comment by jkrasting** [comment]
[00010101.ocean_month_z_refined.nc.txt](https://github.com/NOAA-GFDL/MOM6-examples/files/1360388/00010101.ocean_month_z_refined.nc.txt)
[00010101.ocean_month_refined.nc.txt](https://github.com/NOAA-GFDL/MOM6-examples/files/1360389/00010101.ocean_month_refined.nc.txt)
[00010101.ocean_month_rho2_refined.nc.txt](https://github.com/NOAA-GFDL/MOM6-examples/files/1360390/00010101.ocean_month_rho2_refined.nc.txt)

**Comment by adcroft** [comment]
(I will not explain how I managed to expand those attachments on my phone).

I think the following need to be corrected:
- `wmo:cell_methods = "z_i:point xh:sum yh:sum time:mean" ;`
- Full path in global :filename ?

```
netcdf 00010101.ocean_month_z_refined {
dimensions:
	time = UNLIMITED ; // (12 currently)
	basin = 3 ;
	strait = 17 ;
	strlen = 31 ;
	xh = 1440 ;
	yh = 1080 ;
	yq = 1080 ;
	z_l = 35 ;
	z_i = 36 ;
	nv = 2 ;
variables:
	double time(time) ;
		time:long_name = "time" ;
		time:units = "days since 0001-01-01 00:00:00" ;
		time:cartesian_axis = "T" ;
		time:calendar_type = "NOLEAP" ;
		time:calendar = "NOLEAP" ;
		time:bounds = "time_bnds" ;
	double xh(xh) ;
		xh:long_name = "h point nominal longitude" ;
		xh:units = "degrees_east" ;
		xh:cartesian_axis = "X" ;
	double yh(yh) ;
		yh:long_name = "h point nominal latitude" ;
		yh:units = "degrees_north" ;
		yh:cartesian_axis = "Y" ;
	double yq(yq) ;
		yq:long_name = "q point nominal latitude" ;
		yq:units = "degrees_north" ;
		yq:cartesian_axis = "Y" ;
	char region(basin, strlen) ;
		region:standard_name = "region" ;
	char strait(strait, strlen) ;
		strait:standard_name = "region" ;
	double z_l(z_l) ;
		z_l:long_name = "Depth at cell center" ;
		z_l:units = "meters" ;
		z_l:cartesian_axis = "Z" ;
		z_l:positive = "down" ;
		z_l:edges = "z_i" ;
	double z_i(z_i) ;
		z_i:long_name = "Depth at interface" ;
		z_i:units = "meters" ;
		z_i:cartesian_axis = "Z" ;
		z_i:positive = "down" ;
	double nv(nv) ;
		nv:long_name = "vertex number" ;
		nv:units = "none" ;
		nv:cartesian_axis = "N" ;
	float msftyyz(time, basin, z_i, yq) ;
		msftyyz:_FillValue = 1.e+20f ;
		msftyyz:missing_value = 1.e+20f ;
		msftyyz:time_avg_info = "average_T1,average_T2,average_DT" ;
		msftyyz:long_name = "Ocean Y Overturning Mass Streamfunction" ;
		msftyyz:standard_name = "ocean_y_overturning_mass_streamfunction" ;
		msftyyz:cell_methods = "z_i:sum yq:sum time:mean" ;
		msftyyz:coordinates = "region" ;
		msftyyz:units = "kg s-1" ;
	float msftyzsmpa(time, basin, z_i, yq) ;
		msftyzsmpa:_FillValue = 1.e+20f ;
		msftyzsmpa:missing_value = 1.e+20f ;
		msftyzsmpa:time_avg_info = "average_T1,average_T2,average_DT" ;
		msftyzsmpa:long_name = "ocean Y overturning mass streamfunction due to parameterized submesoscale advection" ;
		msftyzsmpa:standard_name = "ocean_meridional_overturning_mass_streamfunction_due_to_parameterized_submesoscale_advection" ;
		msftyzsmpa:cell_methods = "z_i:sum yq:sum time:mean" ;
		msftyzsmpa:coordinates = "region" ;
		msftyzsmpa:units = "kg s-1" ;
	float msftyzmpa(time, basin, z_i, yq) ;
		msftyzmpa:_FillValue = 1.e+20f ;
		msftyzmpa:missing_value = 1.e+20f ;
		msftyzmpa:time_avg_info = "average_T1,average_T2,average_DT" ;
		msftyzmpa:long_name = "ocean Y overturning mass streamfunction due to parameterized mesoscale advection" ;
		msftyzmpa:standard_name = "ocean_y_overturning_mass_streamfunction_due_to_parameterized_mesoscale_advection" ;
		msftyzmpa:cell_methods = "z_i:sum yq:sum time:mean" ;
		msftyzmpa:coordinates = "region" ;
		msftyzmpa:units = "kg s-1" ;
	float wmo(time, z_i, yh, xh) ;
		wmo:_FillValue = 1.e+20f ;
		wmo:missing_value = 1.e+20f ;
		wmo:time_avg_info = "average_T1,average_T2,average_DT" ;
		wmo:long_name = "Upward mass transport from resolved and parameterized advective transport" ;
		wmo:standard_name = "upward_ocean_mass_transport" ;
		wmo:cell_methods = "z_i:sum xh:sum yh:sum time:mean" ;
		wmo:units = "kg s-1" ;
	float mfo(time, strait) ;
		mfo:_FillValue = 1.e+20f ;
		mfo:missing_value = 1.e+20f ;
		mfo:time_avg_info = "average_T1,average_T2,average_DT" ;
		mfo:long_name = "Sea Water Transport" ;
		mfo:standard_name = "sea_water_transport_across_line" ;
		mfo:cell_methods = "time:mean" ;
		mfo:coordinates = "strait" ;
		mfo:units = "kg s-1" ;
	double average_T1(time) ;
		average_T1:long_name = "Start time for average period" ;
		average_T1:units = "days since 0001-01-01 00:00:00" ;
		average_T1:missing_value = 1.e+20 ;
		average_T1:_FillValue = 1.e+20 ;
	double average_T2(time) ;
		average_T2:long_name = "End time for average period" ;
		average_T2:units = "days since 0001-01-01 00:00:00" ;
		average_T2:missing_value = 1.e+20 ;
		average_T2:_FillValue = 1.e+20 ;
	double average_DT(time) ;
		average_DT:long_name = "Length of average period" ;
		average_DT:units = "days" ;
		average_DT:missing_value = 1.e+20 ;
		average_DT:_FillValue = 1.e+20 ;
	double time_bnds(time, nv) ;
		time_bnds:long_name = "time axis boundaries" ;
		time_bnds:units = "days" ;
		time_bnds:missing_value = 1.e+20 ;
		time_bnds:_FillValue = 1.e+20 ;

// global attributes:
		:filename = "/vftmp/John.Krasting/pid29605/mdt/20171002-dev/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15/gfdl.ncrc4-intel16-prod-openmp/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15_00010101/history_refineDiag/00010101.nc/00010101.ocean_month_z_refined.nc" ;
		:title = "CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15" ;
		:associated_files = "areacello: 00010101.ocean_static.nc" ;
		:grid_type = "regular" ;
		:grid_tile = "N/A" ;
}
```

**Comment by adcroft** [comment]
```
netcdf 00010101.ocean_month_refined {
dimensions:
	time = UNLIMITED ; // (12 currently)
	basin = 3 ;
	strlen = 21 ;
	yq = 1080 ;
	xh = 1440 ;
	nv = 2 ;
variables:
	double time(time) ;
		time:long_name = "time" ;
		time:units = "days since 0001-01-01 00:00:00" ;
		time:cartesian_axis = "T" ;
		time:calendar_type = "NOLEAP" ;
		time:calendar = "NOLEAP" ;
		time:bounds = "time_bnds" ;
	double yq(yq) ;
		yq:long_name = "q point nominal latitude" ;
		yq:units = "degrees_north" ;
		yq:cartesian_axis = "Y" ;
	char region(basin, strlen) ;
		region:standard_name = "region" ;
	double xh(xh) ;
		xh:long_name = "h point nominal longitude" ;
		xh:units = "degrees_east" ;
		xh:cartesian_axis = "X" ;
	double nv(nv) ;
		nv:long_name = "vertex number" ;
		nv:units = "none" ;
		nv:cartesian_axis = "N" ;
	float hfy(time, yq, xh) ;
		hfy:_FillValue = 1.e+20f ;
		hfy:missing_value = 1.e+20f ;
		hfy:time_avg_info = "average_T1,average_T2,average_DT" ;
		hfy:long_name = "Ocean Heat Y Transport" ;
		hfy:standard_name = "ocean_heat_y_transport" ;
		hfy:cell_methods = "yq:point xh:mean time:mean" ;
		hfy:units = "W" ;
	float hfx(time, yq, xh) ;
		hfx:_FillValue = 1.e+20f ;
		hfx:missing_value = 1.e+20f ;
		hfx:time_avg_info = "average_T1,average_T2,average_DT" ;
		hfx:long_name = "Ocean Heat X Transport" ;
		hfx:standard_name = "ocean_heat_x_transport" ;
		hfx:cell_methods = "yh:mean xq:point time:mean" ;
		hfx:units = "W" ;
	float hfbasin(time, basin, yq) ;
		hfbasin:_FillValue = 1.e+20f ;
		hfbasin:missing_value = 1.e+20f ;
		hfbasin:comment = "Indo-Pacific heat transport begins at 34 S" ;
		hfbasin:time_avg_info = "average_T1,average_T2,average_DT" ;
		hfbasin:long_name = "Northward Ocean Heat Transport" ;
		hfbasin:standard_name = "northward_ocean_heat_transport" ;
		hfbasin:cell_methods = "yq:sum time:mean" ;
		hfbasin:coordinates = "region" ;
		hfbasin:units = "W" ;
	double average_T1(time) ;
		average_T1:long_name = "Start time for average period" ;
		average_T1:units = "days since 0001-01-01 00:00:00" ;
		average_T1:missing_value = 1.e+20 ;
		average_T1:_FillValue = 1.e+20 ;
	double average_T2(time) ;
		average_T2:long_name = "End time for average period" ;
		average_T2:units = "days since 0001-01-01 00:00:00" ;
		average_T2:missing_value = 1.e+20 ;
		average_T2:_FillValue = 1.e+20 ;
	double average_DT(time) ;
		average_DT:long_name = "Length of average period" ;
		average_DT:units = "days" ;
		average_DT:missing_value = 1.e+20 ;
		average_DT:_FillValue = 1.e+20 ;
	double time_bnds(time, nv) ;
		time_bnds:long_name = "time axis boundaries" ;
		time_bnds:units = "days" ;
		time_bnds:missing_value = 1.e+20 ;
		time_bnds:_FillValue = 1.e+20 ;

// global attributes:
		:filename = "/vftmp/John.Krasting/pid29605/mdt/20171002-dev/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15/gfdl.ncrc4-intel16-prod-openmp/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15_00010101/history_refineDiag/00010101.nc/00010101.ocean_month_refined.nc" ;
		:title = "CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15" ;
		:associated_files = "areacello: 00010101.ocean_static.nc" ;
		:grid_type = "regular" ;
		:grid_tile = "N/A" ;
}
```

**Comment by adcroft** [comment]
That last file (ocean_month_refined) needs the following corrections:
- `hfbasin:cell_methods = "yq:point time:mean" ;`
- `global:associated_files` is not needed since areacello is not referenced (in ocean_month_refined)

**Comment by adcroft** [comment]
Also, in ocean_month_z_refined:
- `wmo:cell_measures: "area:areacello"`

The associated_files is then correct in ocean_month_z_refined, but still not needed in ocean_month_refined.

**Comment by adcroft** [comment]
Corrections for 
- `rho2_l:standard_name = "sea_water_potential_density"` (see http://clipc-services.ceda.ac.uk/dreq/u/dim:rho.html)
- `msftyrho:cell_methods = "rho2_i:point yq:point time:mean" ;`
- `msftyrhompa:cell_methods = "rho2_i:point yq:point time:mean" ;`
- global:associated_files is not needed because areacello is not referenced.
```
netcdf 00010101.ocean_month_rho2_refined {
dimensions:
	time = UNLIMITED ; // (12 currently)
	basin = 3 ;
	strlen = 21 ;
	yq = 1080 ;
	rho2_l = 35 ;
	rho2_i = 36 ;
	nv = 2 ;
variables:
	double time(time) ;
		time:long_name = "time" ;
		time:units = "days since 0001-01-01 00:00:00" ;
		time:cartesian_axis = "T" ;
		time:calendar_type = "NOLEAP" ;
		time:calendar = "NOLEAP" ;
		time:bounds = "time_bnds" ;
	double yq(yq) ;
		yq:long_name = "q point nominal latitude" ;
		yq:units = "degrees_north" ;
		yq:cartesian_axis = "Y" ;
	char region(basin, strlen) ;
		region:standard_name = "region" ;
	double rho2_l(rho2_l) ;
		rho2_l:long_name = "Target Potential Density at cell center" ;
		rho2_l:units = "kg m-3" ;
		rho2_l:cartesian_axis = "Z" ;
		rho2_l:positive = "down" ;
		rho2_l:edges = "rho2_i" ;
	double rho2_i(rho2_i) ;
		rho2_i:long_name = "Target Potential Density at interface" ;
		rho2_i:units = "kg m-3" ;
		rho2_i:cartesian_axis = "Z" ;
		rho2_i:positive = "down" ;
	double nv(nv) ;
		nv:long_name = "vertex number" ;
		nv:units = "none" ;
		nv:cartesian_axis = "N" ;
	float msftyrho(time, basin, rho2_i, yq) ;
		msftyrho:_FillValue = 1.e+20f ;
		msftyrho:missing_value = 1.e+20f ;
		msftyrho:time_avg_info = "average_T1,average_T2,average_DT" ;
		msftyrho:long_name = "Ocean Y Overturning Mass Streamfunction" ;
		msftyrho:standard_name = "ocean_y_overturning_mass_streamfunction" ;
		msftyrho:cell_methods = "rho2_i:sum yq:sum time:mean" ;
		msftyrho:coordinates = "region" ;
		msftyrho:units = "kg s-1" ;
	float msftyrhompa(time, basin, rho2_i, yq) ;
		msftyrhompa:_FillValue = 1.e+20f ;
		msftyrhompa:missing_value = 1.e+20f ;
		msftyrhompa:time_avg_info = "average_T1,average_T2,average_DT" ;
		msftyrhompa:long_name = "ocean Y overturning mass streamfunction due to parameterized mesoscale advection" ;
		msftyrhompa:standard_name = "ocean_y_overturning_mass_streamfunction_due_to_parameterized_mesoscale_advection" ;
		msftyrhompa:cell_methods = "rho2_i:sum yq:sum time:mean" ;
		msftyrhompa:coordinates = "region" ;
		msftyrhompa:units = "kg s-1" ;
	double average_T1(time) ;
		average_T1:long_name = "Start time for average period" ;
		average_T1:units = "days since 0001-01-01 00:00:00" ;
		average_T1:missing_value = 1.e+20 ;
		average_T1:_FillValue = 1.e+20 ;
	double average_T2(time) ;
		average_T2:long_name = "End time for average period" ;
		average_T2:units = "days since 0001-01-01 00:00:00" ;
		average_T2:missing_value = 1.e+20 ;
		average_T2:_FillValue = 1.e+20 ;
	double average_DT(time) ;
		average_DT:long_name = "Length of average period" ;
		average_DT:units = "days" ;
		average_DT:missing_value = 1.e+20 ;
		average_DT:_FillValue = 1.e+20 ;
	double time_bnds(time, nv) ;
		time_bnds:long_name = "time axis boundaries" ;
		time_bnds:units = "days" ;
		time_bnds:missing_value = 1.e+20 ;
		time_bnds:_FillValue = 1.e+20 ;

// global attributes:
		:filename = "/vftmp/John.Krasting/pid29605/mdt/20171002-dev/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15/gfdl.ncrc4-intel16-prod-openmp/CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15_00010101/history_refineDiag/00010101.nc/00010101.ocean_month_rho2_refined.nc" ;
		:title = "CM4_c96L33_am4p0_2010_30d15_tlt_geo_kd15" ;
		:associated_files = "areacello: 00010101.ocean_static.nc" ;
		:grid_type = "regular" ;
		:grid_tile = "N/A" ;
}
```

**Comment by jkrasting** [comment]
I'm inheriting this from the MOM6 output.  It would be better to make that
change in the MOM6 code (I assume where the coordinate for remapping is
defined).  @aadcroft, @ashao, could either of you make that change?

rho2_l:standard_name = "sea_water_potential_density" (see
http://clipc-services.ceda.ac.uk/dreq/u/dim:rho.html

I'll handle everything else.  Thanks for the QC!

**Comment by jkrasting** [comment]
If msftyrho:cell_methods = "rho2_i:point yq:point time:mean",

Shouldn't msftyyz.cell_methods = 'z_i:point yq:point time:mean' ?

Currently have:  msftyyz.cell_methods = 'z_i:sum yq:sum time:mean' ?

-----
Dr. John Krasting, Physical Scientist
Biogeochemistry, Ecosystems, and Climate Group
NOAA/Geophysical Fluid Dynamics Laboratory
201 Forrestal Road, Princeton, NJ 08540
P. (609) 452-5359   |   F. (609) 987-5063

**Comment by adcroft** [comment]
Also for ocean_month_z_refined:
- `msftyyz:cell_methods = "z_i:point yq:point time:mean" ;`
- `msftyzsmpa:cell_methods = "z_i:point yq:point time:mean" ;`
- `msftyzmpa:cell_methods = "z_i:point yq:point time:mean" ;`

I think it's right that a stream function is `:point` in both directions because you can only interpolate it. The 3d lateral transports are :sum in the tangential directions because they have to be summed when combining, not interpolated.

**Comment by adcroft** [comment]
Sorry @jkrasting , just saw your comment saying what I just did.

**Pull request merged**